### PR TITLE
Add `rawtxwithfee` ZMQ topic handling

### PR DESCRIPTION
**Pull request by 0xB10C**
This PR adds the handling of the `rawtxwithfee` ZMQ topic. Messages with this topic are published by a customized Bitcoin Core instance. Next to a lot of refactors and cleanups, the integration test environment setup is moved to the Makefile. A `bitcoind` instance running in a Docker container is used now. Since the GitHub Actions CI seems to be quite unstable and error prone I moved to Travis CI.

### CODE REVIEW

File: src/zmqsubscriber/zmqsubscriber.go [hunk 1]
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
...
+		z.IncomingBlocks <- *block
+	default:
+		return fmt.Errorf("unknown topic %s", topic)
+	}
+
+	return nil
+}
+
+// Stop sets the cancel flag to true. The ZMQSubscriber is stopped after it finishes
+// receiving a message or
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: `or reaches the timeout`

File: src/zmqsubscriber/zmqsubscriber.go [hunk 2]
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
...
+
+	return &types.Transaction{
+		FirstSeen: firstSeen,
+		TxID:      txid,
+	}, nil
+}
+
+func parseTransaction(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: unbalanced parens

File: src/zmqsubscriber/zmqsubscriber.go [hunk 3]
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
...
+	}
+
+	var txid types.Hash32
+	wireTxHash := wireTx.TxHash()
+	copy(txid[:], []byte(wireTxHash.CloneBytes()))
+
+	fee := binary.BigEndian.Uint64(feeBytes)
+
+	vsizeFloat := (float64(wireTx.SerializeSizeStripped()*3) + float64(wireTx.SerializeSize())) / 4
+	vsize := int(math.Ceil(vsizeFloat))
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: how about we track the weight units instead and keep full precision?

File: src/zmqsubscriber/zmqsubscriber.go [hunk 4]
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
...
+	// which is not used here, but noted for completeness.
+	// rawtxwithfee is the rawtx by Bitcoin Core concatinated with the 8 byte BE
+	// transaction fee
+	rawtxwithfee, _ := payload[0], payload[1]
+
+	length := len(rawtxwithfee)
+	if length <= 8 {
+		return nil, errors.New("unexpected rawtxwithfee length")
+	}
+	rawtx, feeBytes := rawtxwithfee[:length-8], rawtxwithfee[length-8:]
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: I'm not familiar with how it is serialized in bitcoin core, but can we not add this as a separate `payload` element?

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 5]
```diff
@@ -52,32 +52,54 @@ func getTestZMQSubscriber(t *testing.T, env test.TestEnv) *ZMQSubscriber {
...
+	}
 	defer env.Quit()
 
+	addressSendFrom, err := env.GetNewAddress()
+	if err != nil {
+		t.Errorf("could not generate a new bitcoin address: %s", err)
+	}
+
+	// Generate 101 blocks to have spendable UTXOs.
+	env.GenerateToAddress(101, addressSendFrom)
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: I'm ok with removing `panic()` for recoverable errors, but we have to do the err checks then

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 6]
```diff
@@ -52,32 +52,54 @@ func getTestZMQSubscriber(t *testing.T, env test.TestEnv) *ZMQSubscriber {
...
 	require.NotNil(t, waitForZMQBlock(t, z, waitTimeout))
 
-	env.GenerateBlocks(1)
+	env.SendSimpleTransaction(addressSendTo)
 	require.NotNil(t, waitForZMQTransaction(t, z, waitTimeout))
+
+	env.GenerateToAddress(1, addressSendFrom)
 	require.NotNil(t, waitForZMQBlock(t, z, waitTimeout))
 
+	env.SendSimpleTransaction(addressSendTo)
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: err check (I'm ok with `require.NoError`)

File: src/zmqsubscriber/zmqsubscriber.go (hunk 3 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Good point!

**Comment by 0xB10C** [comment]
Thanks for the review. Will address this after I'm happy with the integration tests.

File: src/test/testenv.go [hunk 7]
```diff
@@ -78,61 +81,128 @@ func newBitcoindRpc() (*rpcclient.Client, error) {
...
+	return &TestEnv{
+		cmd:     cmd,
+		rpc:     rpc,
+		ZMQHost: zmqHost,
+		ZMQPort: zmqPort,
+	}, nil
+}
+
+// GetNewAddress returns a new address via the `getnewaddress` RPC.
+func (e *TestEnv) GetNewAddress() (btcutil.Address, error) {
```
[the remark this answers is not in the document]
Comment (0xB10C) [reply src/test/testenv.go]: Somehow didn't see that... Thanks!

File: src/zmqsubscriber/zmqsubscriber.go (hunk 4 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Do you mean a struct?

```go
type payload struct {
	rawtx    []byte,
	feeBytes [8]byte,
}
```

File: src/zmqsubscriber/zmqsubscriber.go (hunk 3 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Changed in https://github.com/0xB10C/bademeister-go/pull/19/commits/08d2148b10a7831418f37b30d78a5e1ae872d65d

**Comment by 0xB10C** [comment]
I think this is finally ready for review after a long battle with Github Action CI and Travis CI. I've updated the initial description to summarize what I've changed.

File: src/bitcoinrpcclient/bitcoinrpcclient.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+	client := &BitcoinRPCClient{rpc}
+
+	return client, nil
+}
+
+// WaitTillRPCServerReady tries every `checkInterval` if the RPC Server the
+// client is connected to is ready. It returns nil if the server is ready, or
+// otherwise times out after `timeout`.
+func (rpcClient *BitcoinRPCClient) WaitTillRPCServerReady(timeout time.Duration) error {
```
Comment (OttoAllmendinger) [inline src/bitcoinrpcclient/bitcoinrpcclient.go]: why not call this in the constructor instead? less work for the user

File: src/storage/storage.go [hunk 9]
```diff
@@ -37,37 +39,54 @@ func NewStorage(path string) (*Storage, error) {
...
+
+	const fillConfigTable string = `
+	INSERT INTO config (version) VALUES (?)
+	`
+
+	if _, err := s.db.Exec(fillConfigTable, version); err != nil {
+		return fmt.Errorf("could not fill the `config` table: %s", err)
+	}
+
+	const crateMempoolTxTable string = `
```
Comment (OttoAllmendinger) [inline src/storage/storage.go]: `create`

File: src/storage/storage.go [hunk 10]
```diff
@@ -130,16 +151,22 @@ func (i *TxIterator) Next() *types.Transaction {
...
 	}
+
 	var txidBytes []byte
 	var firstSeen int64
-	if err := i.rows.Scan(&txidBytes, &firstSeen); err != nil {
-		panic(err)
+	var fee uint64
+	var weight int
+	if err := i.rows.Scan(&txidBytes, &firstSeen, &fee, &weight); err != nil {
+		panic(err) // TODO: cleanup
```
Comment (OttoAllmendinger) [inline src/storage/storage.go]: IMO `panic()` is fine here since the caller can only propagate anyway

File: src/zmqsubscriber/zmqsubscriber.go [hunk 11]
```diff
@@ -155,6 +118,98 @@ func (z *ZMQSubscriber) Run() error {
...
+	// TODO: use GetTime() and allow other time sources (eg NTP-corrected)
+	firstSeen := time.Now().UTC()
+
+	switch topic {
+	case topicHashTx:
+		tx, err := parseTransactionHash(firstSeen, payload)
+		if err != nil {
+			return err
+		}
+		z.IncomingTx <- *tx
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: we can prob get rid of that later

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 12]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+	cookieExists := false
+	currentTimeout := 0 * time.Second
+
+	fmt.Printf("Waiting for '%s'...\n", cookiePath)
+
+	for !cookieExists {
+
+		if currentTimeout > maxTimeout {
+			fmt.Printf("Timeout (%s) while waiting for '%s'.\n", maxTimeout, cookiePath)
+			os.Exit(1)
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: `error` return would be a bit more standard

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 13]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+	rpcClient.GenerateToAddress(101, addressMineTo)
 
-	env := test.NewTestEnv()
-	defer env.Quit()
+	addressSendTo, err := rpcClient.GetNewAddress("addressSendTo")
+	require.NoError(t, err)
 
-	z := getTestZMQSubscriber(t, env)
+	z, err := setupAndRunZMQSubscriber(t, zmqHost, zmqPort)
 	defer z.Stop()
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: should come after `require.NoError`

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 14]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+	require.NoError(t, err)
+	require.NotNil(t, waitForZMQTransaction(t, z, zmqWaitTimeout))
 
 	// Test second subscriber.
 	// This demonstrates that it should be possible to connect multiples instance of
 	// Bademeisterd and have continuous capture.
 	{
-		z2 := getTestZMQSubscriber(t, env)
+		z2, err := setupAndRunZMQSubscriber(t, zmqHost, zmqPort)
 		defer z2.Stop()
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: after `require.NoError()`

File: src/zmqsubscriber/zmqsubscriber.go [hunk 15]
```diff
@@ -59,92 +46,68 @@ func NewZMQSubscriber(host string, port string) (*ZMQSubscriber, error) {
...
+			} else if err == zmq4.Errno(syscall.EINTR) {
 				continue
 			}
-
-			return fmt.Errorf("Could not receive ZMQ message: %s %v", err, err)
+			return fmt.Errorf("could not receive ZMQ message: %s", err)
 		}
 
 		topic, payload := string(msg[0]), msg[1:]
-		log.Printf(`[ZMQ] received topic "%s"`, topic)
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: the idea here was to have prefixes for different subsystems that can be easily grepped

File: src/zmqsubscriber/zmqsubscriber.go (hunk 2 above)
Comment (OttoAllmendinger) [reply src/zmqsubscriber/zmqsubscriber.go]: still unbalanced ;)

File: src/zmqsubscriber/zmqsubscriber.go (hunk 4 above)
Comment (OttoAllmendinger) [reply src/zmqsubscriber/zmqsubscriber.go]: I meant putting it between `payload[0]` and `payload[1]` but we clarified out-of-band

File: src/storage/storage.go [hunk 16]
```diff
@@ -3,15 +3,17 @@ package storage
...
+	"log"
 	"os"
 	"time"
 
 	"github.com/0xb10c/bademeister-go/src/types"
 	_ "github.com/mattn/go-sqlite3"
 )
 
-const VERSION = 2
+const currentVersion = 4
```
Comment (OttoAllmendinger) [inline src/storage/storage.go]: IMO it is fine to bump more than 1 version per PR but not required

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 17]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+	}
 }
 
 func TestZMQSubscriber(t *testing.T) {
-	const waitTimeout = 5 * time.Second
+	if testing.Short() {
+		t.Skip("Skipping " + t.Name() + " since it's not a unit test.")
+	}
+
+	const zmqWaitTimeout = 5 * time.Second
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: nit: if always we pass the same timeout anyway (and don't want to make it a global const either) we can make it a variable that is local to the func

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 18]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+			cookieExists = true
+			break
+		}
+	}
 }
 
 func TestZMQSubscriber(t *testing.T) {
-	const waitTimeout = 5 * time.Second
+	if testing.Short() {
+		t.Skip("Skipping " + t.Name() + " since it's not a unit test.")
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: I'm a bit surprised that we don't have to return here, but it looks correct https://golang.org/pkg/testing/#hdr-Skipping

does it call `os.Exit()` then?

File: src/storage/storage_test.go [hunk 19]
```diff
@@ -44,16 +43,21 @@ func testQueryTransactions(t *testing.T, st *Storage, firstSeen time.Time, txs [
...
-		if !os.IsNotExist(err) {
-			t.Fail()
-		}
+func TestStorage(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping " + t.Name() + " since it's not a unit test.")
 	}
 
-	// create from empty file
+	// The environment variable `TEST_INTEGRATION_DIR` is set to a temporay directory created by
```
Comment (OttoAllmendinger) [inline src/storage/storage_test.go]: temporary

File: src/zmqsubscriber/zmqsubscriber.go [hunk 20]
```diff
@@ -155,6 +118,98 @@ func (z *ZMQSubscriber) Run() error {
...
+func parseTransaction(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
+	}
+
+	// payload[1] contains a 16bit LE sequence number provided by Bitcoin Core,
+	// which is not used here, but noted for completeness.
+	// rawtxwithfee is the rawtx by Bitcoin Core concatinated with the 8 byte BE
+	// transaction fee
+	rawtxwithfee, _ := payload[0], payload[1]
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: pls add a link to the source

File: src/zmqsubscriber/zmqsubscriber_test.go [hunk 21]
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
...
+	return z, nil
+}
+
+// waitForBitcoinRPCCookie blocks while waiting for the Bitcoin Core RPC
+// `.cookie` file to be present. Times out after `maxTimeout`.
+func waitForBitcoinRPCCookie() {
+	const maxTimeout = 10 * time.Second
+	const timeoutDelay = 250 * time.Millisecond
+
+	// The environment variable `TEST_INTEGRATION_DIR` is set to a temporay directory created by
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber_test.go]: `temporary`

File: src/storage/storage.go (hunk 10 above)
Comment (0xB10C) [reply src/storage/storage.go]: Agree, will remove it.

File: src/zmqsubscriber/zmqsubscriber.go (hunk 15 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Could use https://golang.org/pkg/log/#SetPrefix or set the `Lshortfile` flag (https://golang.org/pkg/log/#pkg-constants) for a short filename in the log.

**Comment by OttoAllmendinger** [comment]
Please consider #20 instead of providing the custom bitcoind binary for tests. We shouldn't ask potential contributors to run opaque binaries.

**Comment by 0xB10C** [comment]
Agree. Will have a look at building a docker image with the patched bitcoind next. Is that in the scope of this PR?

**Comment by OttoAllmendinger** [comment]
I think it should be a different PR, but if we can get it in before this one we can avoid having the binary in the git history

File: src/zmqsubscriber/zmqsubscriber_test.go (hunk 18 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber_test.go]: Tests are run in goroutines. `Skip()` exits the current goroutine and marks the tests as "skipped".

File: src/storage/storage.go (hunk 16 above)
Comment (0xB10C) [reply src/storage/storage.go]: Agree for future PRs, won't change it back now..

File: src/zmqsubscriber/zmqsubscriber_test.go (hunk 17 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber_test.go]: I think that's what it is now, isn't it? It's just `const` because it should not be modified.

**Comment by 0xB10C** [comment]
Addressed some of the outstanding comments.

File: src/zmqsubscriber/zmqsubscriber_test.go (hunk 18 above)
Comment (OttoAllmendinger) [reply src/zmqsubscriber/zmqsubscriber_test.go]: but how does `t.Skip()` exit the parent goroutine?

Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber_test.go]: with `runtime.Goexit()`. See https://golang.org/pkg/testing/#B.SkipNow

Comment (OttoAllmendinger) [reply src/zmqsubscriber/zmqsubscriber_test.go]: TIL!

File: src/zmqsubscriber/zmqsubscriber.go [hunk 22]
```diff
@@ -155,6 +118,99 @@ func (z *ZMQSubscriber) Run() error {
...
+
+// Stop sets the cancel flag to true. The ZMQSubscriber is stopped after it
+// finishes receiving a message or reaches the timeout.
 func (z *ZMQSubscriber) Stop() {
 	z.cancel = true
 }
+
+func parseTransactionHash(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
```
Comment (OttoAllmendinger) [inline src/zmqsubscriber/zmqsubscriber.go]: `expected len([txhash, sequence]) == 2`

File: src/zmqsubscriber/zmqsubscriber_test.go (hunk 17 above)
Comment (OttoAllmendinger) [reply src/zmqsubscriber/zmqsubscriber_test.go]: I meant local to `waitForForZMQBlock`

File: src/zmqsubscriber/zmqsubscriber.go (hunk 22 above)
Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Maybe I keep fixing up the wrong commit? Changed this three times now...

Comment (0xB10C) [reply src/zmqsubscriber/zmqsubscriber.go]: Oh I see. It's in `parseTransactionHash`. Going to remove that one in an extra commit.

**Comment by 0xB10C** [comment]
Removed the `bitcoind` binary. Now uses a patched `bitcoind` running inside a Docker container based on this image https://hub.docker.com/r/b10c/bitcoind-patched-zmq.

**Pull request merged**